allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -213,92 +163,103 @@ def read(self, file_path: str) -> Dataset:
                 "did you forget to call the superclass constructor?"
             )
 
+        if lazy:
+            return AllennlpLazyDataset(self.instance_iterator, file_path)
+
+        instances = list(self.instance_iterator(file_path))
+        if not instances:
+            raise ConfigurationError(
+                "No instances were read from the given filepath {}. "
+                "Is the path correct?".format(file_path)
+            )
+
+        return AllennlpDataset(instances)
+
+    def instance_iterator(self, file_path: str) -> Iterable[Instance]:
+        cache_file: Optional[str] = None
         if self._cache_directory:
             cache_file = self._get_cache_location_for_file_path(file_path)
-        else:
-            cache_file = None
 
-        if lazy:
-            lazy_instances: AllennlpLazyDataset = _LazyInstances(
-                self._read,
-                file_path,
-                cache_file,
-                self.deserialize_instance,
-                self.serialize_instance,
-            )
-            if self.max_instances is not None:
-                lazy_instances = _MaxLazyInstances(lazy_instances, self.max_instances)
-            if not self.manual_distributed_sharding and util.is_distributed():
-                lazy_instances = _DistributedLazyInstances(lazy_instances)
-            return lazy_instances
-        else:
-            # First we read the instances, either from a cache or from the original file.
-            if cache_file and os.path.exists(cache_file):
-                instances = self._instances_from_cache_file(cache_file)
-            else:
-                instances = self._read(file_path)
-
-            if self.max_instances is not None:
-                if isinstance(instances, list):
-                    instances = instances[: self.max_instances]
-                else:
-                    instances = itertools.islice(instances, 0, self.max_instances)
-            if not self.manual_distributed_sharding and util.is_distributed():
-                from torch import distributed
-
-                logger.info(
-                    "Returning instances i%%%d==%d",
-                    distributed.get_world_size(),
-                    distributed.get_rank(),
+        if cache_file is not None:
```

Comment:
Wouldn't it be easier (and faster) to have one cache file per worker, which contains only the instances for that one worker? Then you don't need locking or slicing in the cache case. The downside is that you can't re-use the cache when the number of workers changes.